openmls/src/group/core_group/new_from_external_init.rs

```diff
@@ -0,0 +1,148 @@
+
+        if treesync.tree_hash() != verifiable_public_group_state.tree_hash() {
+            return Err(ExternalInitError::TreeHashMismatch);
+        }
+
+        let pgs_signer_leaf = treesync.leaf(verifiable_public_group_state.signer_index())?;
+        let pgs_signer_credential = pgs_signer_leaf
+            .ok_or(ExternalInitError::UnknownSender)?
+            .key_package()
+            .credential();
+        let pgs: PublicGroupState = verifiable_public_group_state
+            .verify(backend, pgs_signer_credential)
+            .map_err(|_| ExternalInitError::InvalidPublicGroupStateSignature)?;
+
+        let (init_secret, kem_output) = InitSecret::from_public_group_state(backend, &pgs)?;
+
+        // We create the GroupContext with the values from the PGS, even though
+        // we already changed the tree by adding our own leaf, thus invalidating
+        // the tree hash.
+        let group_context = GroupContext::new(
+            pgs.group_id,
+            pgs.epoch,
+            pgs.tree_hash.as_slice().to_vec(),
+            pgs.confirmed_transcript_hash.into(),
+            pgs.group_context_extensions.as_slice(),
+        )?;
+
+        // The `EpochSecrets` we create here are essentially zero, with the
+        // exception of the `InitSecret`, which is all we need here for the
+        // external commit.
+        let epoch_secrets = EpochSecrets::with_init_secret(backend, init_secret)?;
+        let (group_epoch_secrets, message_secrets) = epoch_secrets.split_secrets(
+            group_context.tls_serialize_detached()?,
+            treesync.leaf_count()?,
+        );
+
+        // Prepare interim transcript hash
+        let group = CoreGroup {
+            ciphersuite,
+            group_context,
+            tree: treesync,
+            interim_transcript_hash: pgs.interim_transcript_hash.into(),
+            use_ratchet_tree_extension: enable_ratchet_tree_extension,
+            mls_version: pgs.version,
+            group_epoch_secrets,
+            message_secrets,
+        };
+
+        let external_init_proposal = Proposal::ExternalInit(ExternalInitProposal::from(kem_output));
+
+        let mut proposal_store = ProposalStore::default();
+        for proposal in proposals_by_reference {
+            let staged_proposal =
+                StagedProposal::from_mls_plaintext(ciphersuite, backend, proposal.clone())?;
+            proposal_store.add(staged_proposal)
+        }
+
+        let mut inline_proposals = proposals_by_value.to_vec();
+        inline_proposals.push(external_init_proposal);
+
+        let params = CreateCommitParams::builder()
+            .framing_parameters(framing_parameters)
+            .credential_bundle(credential_bundle)
+            .proposal_store(&proposal_store)
+            .inline_proposals(inline_proposals)
+            .commit_type(CommitType::External)
+            // Populate the path
+            .force_self_update(true)
```

Comment:
This should actually happen automatically. Can you adjust the `path_required` flag?